### ISSUE TITLE
Line 442  in  PDCNet.py.

### ISSUE THREAD

**Issue by zwyking** [PruneTruong/DenseMatching#5]
Hello, thanks for your excellent works and codes. Here, i have an issue for the line 442 in PDCNet.py. I think the codes should be "c_t=c23, c_s=c13 ", which means the source is image1 and  target is image2. (Maybe I wrongly understand your code, please correct me (＠_＠;))

**Comment by PruneTruong** [comment]
Hi, so  im1 is the target image (also referred to as the reference image, for example in PDC-Net paper) and im2 is the source image (also referred to as the query image, for example in PDC-Net paper). Therefore, the provided original version of the code is correct. Here, we estimate the flow field relating the target/reference to the source/query. This flow can be used to align the source/query to the target/reference by warping the source/query according to the flow. Let me know if you have other questions :)

**Comment by zwyking** [comment]
Thanks for your explanation for this issue. I have got it and understood this part. I find another problem, that memory consumption will gradually become larger with the training epoch. I want to know if it's normal.

**Comment by PruneTruong** [comment]
I don't think I had this issue before, but let me double check the published version, I will get back to you.

**Comment by zwyking** [comment]
Ok. This phenomenon is obvious in PDCNet-stage2.  During my training precossing, the memory consumption is nearly to 200G with epochs=25.

**Comment by PruneTruong** [comment]
You're right, there seems to be a memory leakage that I didn't have before. I will investigate, in the meanwhile, setting the number of workers to 0 considerably reduces the required memory but might make the training slower. Also, if it crashes, you can restart it, it will restart from where it crashed. Sorry for the inconvenience

**Comment by zwyking** [comment]
Ok, thanks for your patient responses. This is a great work.

**Comment by zwyking** [comment]
I think the memory leakage would occur in the dataloader, because i find the memory add when a new epoch starts.

**Comment by zwyking** [comment]
![QQ截图20210828094629](https://user-images.githubusercontent.com/45560490/131202887-07703cf0-92ed-4c95-802e-b5fec4bde76b.png)
This is a snapshot of PDCNet_stage2 with 6 epochs. It's will be found a clear memory leak. Meanwhile, I notice a big memory increase after a validation.

**Comment by PruneTruong** [comment]
Hi, I am really sorry I was running the code on my main internal repo and I only realized now that i hadn't correctly pushed all files to this repo. Therefore I was not seeing the leak that you see.
It was in https://github.com/PruneTruong/DenseMatching/blob/main/training/losses/multiscale_loss.py,  .item() were forgotten in the logging of the loss, so i was dragging the whole gradient throughout the epochs. Without this, I trained on exactly this repo. 

In Stage 1, I have an increase of about 1.8 GB WITHIN one epoch (reduced to about 1GB if using cv2.imread instead of imageio.imread) but the memory is freed at the end of the epoch. Therefore, i only have an absolute increase of 400MB at the end of first epoch (because creating all the classes for logging metrics) and top 200MB between the epochs after that (due to the logging in memory). 

For stage2, I have an increase of about 5GB within one epoch, also released at the end of the epoch. Absolute increase between the epochs of top 100MB. 

I am very sorry about the inconvenience and thanks for your patience!

**Comment by zwyking** [comment]
yeah, thanks for your reply. I have corrected this problem during my debug. Thanks for your excellent work!
